lib/internal/streams/operators.js

```diff
@@ -0,0 +1,130 @@
+'use strict';
+
+const console = require('console');
+const { AbortController } = require('internal/abort_controller');
+const { AbortError } = require('internal/errors');
+const Readable = require('internal/streams/readable');
+const eos = require('internal/streams/end-of-stream');
+
+module.exports.map = function map(stream, fn, options) {
+  // TODO: Argument validation
+
+  let concurrency = 1;
+  if (Number.isFinite(options)) {
+    concurrency = options;
+  } else if (options && Number.isFinite(options.concurrency)) {
+    concurrency = options.concurrency;
+  }
+
+  let highWaterMark = 1;
+  if (options && Number.isFinite(options.highWaterMark)) {
+    highWaterMark = options.highWaterMark;
+  }
+  highWaterMark = Math.max(0, highWaterMark - concurrency)
+
+  let objectMode = stream.readableObjectMode ?? stream.objectMode ?? true;
+  if (options && typeof options.objectMode === 'boolean') {
+    objectMode = options.objectMode;
+  }
+
+  const ac = new AbortController();
+  const signal = ac.signal;
+  const queue = [];
+
+  let reading = false;
+
+  const ret = new Readable({
+    objectMode,
+    highWaterMark,
+    read () {
+      if (!reading) {
+        read();
+      }
+    },
+    destroy (err, callback) {
+      if (!err && !this.readableEnded) {
+        err = new AbortError();
+      }
+      ac.abort();
+      callback(err);
+    }
+  });
+
+  async function read () {
+    if (ret.readableLength && ret.readableLength >= ret.readableHighWaterMark) {
+      return;
+    }
+
+    try {
+      reading = true;
+      while (queue.length && !ret.destroyed) {
+        let entry = queue.shift();
+
+        if (typeof entry.then === 'function') {
+          entry = await entry;
```

Comment:
Any reason we don't just `await` it anyway? It is fine and a no-op to await non-promises (sans the microtask)

Reply:
To avoid to the micro task...

Reply:
removed for now.